### ISSUE TITLE
Allow user to select which property gets highlighted

### ISSUE THREAD

**Issue by flatsiedatsie** [mozilla-iot/gateway#1826]
As mentioned in [this issue](https://github.com/mozilla-iot/gateway/issues/1809), it would be great if the user can select which property gets to take the 'center stage' spot in a device.

![co2-settings](https://user-images.githubusercontent.com/805405/56756659-c54b4f80-6781-11e9-9605-bb48eff70767.png)

One reason this came up is because of the MySensors add-on. The MySensors Arduino library allows users to create any combination of properties in a new device.

In practise, it was hard to match all MySensors possibilities to existing properties from the Mozilla schema. The result is that 
- most devices don't show data in the central position. They show a generic 'thing' instead, which feels like a waste.
- If a schema can be followed, a user might end up with an undesired highlighted item.

USER STORY
As a user I'd like to manually select which property is highlighted in the things overview. I'd like to make sure there a value is always highlighted, and I avoid the generic 'thing' display.

SOLUTIONS
There is already a dropdown in the thing edit menu. If I'm honest I had no idea what it did, because whenever I switched to another MULTI_LEVEL_SENSOR item in the selector nothing really changed. I was hoping it could be that selector.

Another option that was discussed is to create a new more universal type/property pair that can act as a 'backup' property in case a more perfect choice is not available.

To see the issue in code, have a look at [this file from the adapter](https://github.com/createcandle/Webthings-mysensors-adapter/blob/master/pkg/mysensors_device.py#L92). There you can see that the strict Mozilla schema could only be followed in a small number of cases.

**Comment by flatsiedatsie** [comment]
Here are some sketches:

This sketch shows how a user can select which property should be highlighted:
![select-highlighted-property](https://user-images.githubusercontent.com/805405/56797406-274c9900-6804-11e9-86e1-59fadc09a88b.png)

This sketch is a little more far-out. It explores the idea that users can also select which icon they'd want to represent their device.
![co2-settings-and-icon-select](https://user-images.githubusercontent.com/805405/56797486-53681a00-6804-11e9-8d7c-a254a628f977.png)

**Comment by kgiori** [comment]
I really like this idea. Even for completely custom things, show the list of properties of that thing and let the user pick one to be the default value to display with the icon. What about a thing with actions and events? When would it be valuable for the Edit menu to also include a listing of the things' actions and events?

**Comment by flatsiedatsie** [comment]
Sorry to beat this horse, but it's still a little counter intuitive that this doesn't work.

![gateway_select_highlighted_property](https://user-images.githubusercontent.com/805405/72267666-6338c000-3618-11ea-8c04-67f08c203be0.gif)
_Trying to set the 2.5 measurement of a dust sensor as the highlighted property_

**Comment by flatsiedatsie** [comment]
Perhaps having two multi level sensor `@types` is the cause of the highlighted value being zero. 
![gateway_highlighted_value_zero](https://user-images.githubusercontent.com/805405/72342644-d5ff7500-36c4-11ea-8e73-577bd27ace1b.png)

I now remember the discussion about the capabilities long ago, and I'm probably using it wrong again by having two MultiLevelSensor @types with two properties. I've ended back at the same mistake :-D

**Comment by Nca78** [comment]
I just finished a board for lighting in a room, with multiple relays, each having a switch, and a "master" switch that toggles all the relays on the same board. But the interface "chose" one of the relay switch to use it on main screen and I have no way to change it, this is frustrating.

Is there no way to base the "default" value on the internal index or name of the sensor ? So, in short select a sensor instead of just selecting a type ?
![image](https://user-images.githubusercontent.com/17949850/72603368-75ce3480-394b-11ea-9ea9-70fc485599a5.png)

**Comment by flatsiedatsie** [comment]
As a side note:

@nca78 Since I know you're using the MySensors add-on: if you give your properties a child-ID above 200 it will clone each propoerty into a separate thing too. That way you have things with only one property, which avoids this issue. I built that for voice control, since that way you can give each property a different name.

**Comment by Nca78** [comment]
@flatsiedatsie yes it's a great solution for some cases but not here. I want only one button to switch on/off all the ceiling lights in my living room, and then I can go in the detailed view to switch on/off some areas, it's illogical to split into different switches on the main page and to not have access to all areas + master switch in one single place (detailed view).

With the thing as in the above detailed view, it would be great (and feel more logic imho) to have this kind of selection in the edit view:
![image](https://user-images.githubusercontent.com/17949850/72616964-8f32a900-396a-11ea-9771-1b8b51062807.png)

**Comment by mrstegeman** [comment]
@Nca78 I like your thinking, but it's not always quite so simple. In the case of basic devices, like `OnOffSwitch` and `MultiLevelSensor`, it's straightforward to map individual properties, since the top-level icon is just a view/control of a single property. However, for more advanced devices like `Light`, `MultiLevelSwitch`, or `Thermostat`, the icon is actually a composite of several different properties. Without some specific naming convention, or the use of object properties (#1383), there's no way to tell which properties go with one another.

**Comment by flatsiedatsie** [comment]
But couldn't the dropdown show the amalgamated options as well as just the normal propoerties? Granted, the list would become longer. But I whole heartedly agree withNCA78 that the current situation feels unintuitive.

![combined_select_property](https://user-images.githubusercontent.com/805405/72635202-92bf3380-395c-11ea-8a89-b950fd1a5370.png)

In the bigger picture I feel that the capabilities model is restrictive, because it tries to boil multiple properties into a single small circle in the first place. Things would be much simpler and more flexible if users could select multiple properties that they'd like to have highlighed.

![capability](https://user-images.githubusercontent.com/805405/72636582-a029ed00-395f-11ea-8adc-6469e0c057f5.png)


If capabilities that fuse multiple properties were allowed to be bigger it might be really beneficial:
- Capabilities can still be specially designed for cool integrations that are more than the sum of its parts. But in other cases it could just be a case of taking properties and 'smooshing' them side by side. Users could even be given more contol, and be allowed to select multiple properties that they want highlighted. The interface would hen show their selected properties in this merged-side-by-side way.
BenFrancis has said that he's not married to the octopus design style, so it could be liberating to explore how stepping away from that harness might simplify the backend. Simultaneously, trusting users with a little more freedom to display their system could be welcome for them. Win-win.

- Having the things overview be more of a masonry display, where some items can take more space, would also help with navigation. Gestalt perception principles allow people to quickly sort things by size. Instead of a see of uniform circles, people could instantly spot certain devices by their size/shape. (Heck, I'd allow people to give things a bit of color, that also really helps with scanning).

**Comment by flatsiedatsie** [comment]
People could manage how they want their thing displayed on 'edit thing' page.

If they were also data-logging some properties, they could gain additional special capabilties, such as also selecting to show data from the last hour as a mini graph next to the current value:

![mini-graph](https://user-images.githubusercontent.com/805405/72638170-2b58b200-3963-11ea-88d8-2e180907dd44.png)

Here's a basic amalgamation of three properties that expose everthing the parts you may want direct access to from the radio thing:

![radio](https://user-images.githubusercontent.com/805405/72638402-a6ba6380-3963-11ea-890d-9f165117a47c.png)

All this would help the thing page become more of a useful dashboard. Of course a dashboard could be a separate add-on instead, as was [suggested earlier](https://github.com/mozilla-iot/gateway/issues/1597).

**Comment by flatsiedatsie** [comment]
Just having some fun. This is what the things settings page could look like. 

![thing_settings_page_sketch](https://user-images.githubusercontent.com/805405/72641280-0ca9e980-396a-11ea-9133-e0a1b56a66e7.png)

Users could drag properties from highlighted to 'normal'. Normal being they are visible on the properties overview page.

This idea has an interesting bonus: In the backend the gateway wouldn't really need to technically distinguish between thing settings and properties, perhaps simplifying development even more. For example, all could be handled by the same permissions system. Rules could even respond to settings, etc.

Users get a lot of freedom to manage what aspects of their device they want to have quick access to. If a user wanted to, they could even take a setting item, and drag it all the up into the properties overview, or even make it highlighted. Why not. Vice versa, a user could hide a property they never wanted to change on the settings page. For example if they never wanted to change the station, they could drag that property down to the depths of the settings level.

The only diference between properties and thing settings is that users are not allowed to change the name of settings items. Otherwise this could be abused, or a user might no longer know what a setting is for. Properties are allowed to be renamed (which solved [that issue](https://github.com/mozilla-iot/gateway/issues/1898) too), as their type can always still be distinguished by their icon and value.

**Comment by flatsiedatsie** [comment]
Come to think of it, this page wouldn't need another hamburger icon in the bottom right corner. Please ignore that ;-)

**Comment by Hywan** [comment]
Any updates on this front? :-)

**Comment by flatsiedatsie** [comment]
I've created an addon called Highlights that attemps to address this issue. It allows you to select any property, and then that property will become a new thing which has that property as its highlight.

Simply select a property in the UI:
![highlights_ui](https://user-images.githubusercontent.com/805405/89082221-a2004e80-d38d-11ea-95b0-707b21c53100.png)

Then, go to the things overview page and click (+) to add the new things.

![highlights-things](https://user-images.githubusercontent.com/805405/89082380-205cf080-d38e-11ea-91a0-bfd142c6ae99.png)


If/when the Gateway allows users to select any property as the highlighted property, then this addon will happily be retired. Until then this addon can make it easier to use the things overview page as an overview of your device's current values, without being limited by the capability conventions.
